phoenicis-engines/src/main/java/org/phoenicis/engines/EngineSettingsManager.java

```diff
@@ -19,98 +19,128 @@
 
-        interactiveScriptSession.eval(this.createFetchScript(repositoryDTO),
-                output -> callback.accept(((Value) output).as(Map.class)),
-                errorCallback);
+                                final Value settingClass = (Value) phoenicisScriptEngine.evalAndReturn(include,
+                                        errorCallback);
+
+                                return settingClass.newInstance().as(EngineSetting.class);
+                            }, Collectors.toList())));
+
+            callback.accept(result);
+        });
     }
 
     /**
-     * retrieves a Javascript string which can be used to fetch the available settings
-     * @param repositoryDTO repository containing the settings
-     * @return Javascript
+     * Fetches a list of the setting parameters requires to fetch all settings in the given repository
+     *
+     * @param repositoryDTO The repository containing the settings
+     * @return A list containing all necessary setting parameters
      */
-    private String createFetchScript(RepositoryDTO repositoryDTO) {
+    private List<SettingConfig> fetchSettingConfigurations(RepositoryDTO repositoryDTO) {
         // get engine CategoryDTOs
-        List<CategoryDTO> categoryDTOS = new ArrayList<>();
-        for (TypeDTO typeDTO : repositoryDTO.getTypes()) {
-            if (typeDTO.getId().equals("engines")) {
-                categoryDTOS = typeDTO.getCategories();
-            }
-        }
-        StringBuilder script = new StringBuilder();
-        script.append("(function () {\n");
-        script.append("var settings = {};\n");
-        for (CategoryDTO engine : categoryDTOS) {
-            final String engineId = engine.getId().replaceAll("^.*\\.", "");
-            for (ApplicationDTO applicationDTO : engine.getApplications()) {
-                if (applicationDTO.getId().equals("engines." + engineId + ".settings")) {
-                    for (ScriptDTO scriptDTO : applicationDTO.getScripts()) {
-                        script.append("include(\"engines." + engineId + ".settings."
-                                + scriptDTO.getId().replaceAll("^.*\\.", "") + "\");\n");
-                        script.append("if (!(\"" + engineId + "\" in settings))\n");
-                        script.append("{\n");
-                        script.append("settings[\"" + engineId + "\"] = new java.util.ArrayList();\n");
-                        script.append("}\n");
-                        script.append("settings[\"" + engineId + "\"].add(new Setting());\n");
-                    }
-                }
-            }
+        List<CategoryDTO> categoryDTOs = repositoryDTO.getTypes().stream()
+                .filter(typeDTO -> typeDTO.getId().equals("engines"))
+                .flatMap(typeDTO -> typeDTO.getCategories().stream())
+                .collect(Collectors.toList());
+
+        return categoryDTOs.stream()
```

Comment:
@plata I've replaced the old `createFetchScript` method with this one because it didn't work after changing the scripts. The idea here is to first fetch all valid `engineId`, `settingId` pairs and then fetch the corresponding scripts afterwards in a separate step.

If you don't like this "flat" style I'm also willing to reintroduce a for loop to create the result `List`, what do you think?

Reply:
I don't really care as long as it works. On the long run, I would like to have something like `include(\"engines.%s.settings\")` to return a list of settings.